docs/src/components/menu/Menu.tsx

```diff
@@ -13,10 +15,39 @@ interface Props {
  * prerendered.
  */
 export default function Menu({ navigationItems = [], openAPILinks }: Props) {
+  const router = useRouter();
+  const currentPath: string[] = router.asPath.split('/').slice(1);
+  let currentPageName: string | undefined;
+  let currentpageItems = navigationItems;
+  let currentPagePath = '/';
+
+  /* Get sections of the current page */
+  // Check segments of current route to find the navigationItems
+  currentPath.map((path, index) => {
+    currentpageItems.map((section) => {
+      section.links?.map((link) => {
+        // Check all the pages for link of the route segment we are going over
+        if (link.page && link.to?.split('/')[index + 1] === path) {
+          currentpageItems = link.subpage as SitemapItem[];
+          currentPageName = link.name;
+          currentPagePath = link.to;
+        }
+      });
+    });
+  });
```

Comment:
What's happening here? Three nested maps, but we don't use the map result.

We're better off using `forEach` to make the intent of the loop obvious, but I also wonder if this is the most efficient way. Can this be simplified?

Reply:
good catch! we should definitely be using a `forEach`.
I have refactored the code to be more readable by using a `.filter` and `.find`

```javascript
  // Find and get the current page item
  currentPath.forEach((pathSegment, pathDepthIndex) => {
    currentSubpages.forEach((section) => {
      const currentPageItem = section.links
        ?.filter((link) => link.page === true)
        .find((link) => splitPath(link.to as string)[pathDepthIndex] === pathSegment);

      if (currentPageItem === undefined) return;
      currentSubpages = currentPageItem.subpage as SitemapItem[];
      currentPageName = currentPageItem.name;
      currentPagePath = currentPageItem.to as string;
    });
  });
  ```
  
  Thanks again for the valuable suggestion!

Reply:
I'm not completely sure what the code does, so please feel free to ignore this comment 🙂


I assume that `link.page === true` is connected to `link.to as string`? (BTW `link.page === true` ~= `link.page`). I wonder if we could make the type of link more sophisticated to encode the rule at the type level so that doing `link.page === true` would narrow `link.to` to `string` and not `string | undefined`.

The double nested loop with 3 mutations is a bit hard to read. Have you considered doing something like:

```
const subpagesPaths = something.map(...) // prepare paths to be compared to router.asPath
const currentPageItem = subpagesPaths.find(...)
...
```

Also

```
const currentPageItem = section.links
  ?.filter((link) => link.page === true)
  .find((link) => splitPath(link.to as string)[pathDepthIndex] === pathSegment)
```

could be an early return

```
if (!section.links) return

const currentPageItem = section.links
  .filter((link) => link.page === true)
  .find((link) => splitPath(link.to as string)[pathDepthIndex] === pathSegment)
```

And I guess you could change `currentPageItem.to as string` to `currentPageItem.to || ''` and maybe see if the cast `as SitemapItem[]` is avoidable?

---

Again, please feel free to ignore or push back against my nitpicks! 🙏

Reply:
Thanks @3v0k4! Small addition, if you already do a find, there's no need to filter first, and it's better to ensure we're dealing with a string, then to typecast to it. 

```
if (!section.links) return

const currentPageItem = section.links
  .find((link) => link.page && splitPath(String(link.to))[pathDepthIndex] === pathSegment)
```

Reply:
Thanks @3v0k4 I have changed the code accordingly to your suggestions!

> if we could make the type of link more sophisticated to encode the rule at the type level 
I have also been thinking about this.
Perhaps someway to derive the type for the Sitemap completely from the `sitemap.json` (I have seen TypeScript inferring objects)
It's definitely something worth experimenting around.

@smeijer removed that extra filter and added the `link.page &&`